bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -2382,4 +2383,10 @@ public void closeComplete(int rc, LedgerHandle lh, Object ctx) {
         }
     }
 
+    List<BookieSocketAddress> getCurrentEnsemble() {
+        // Getting current ensemble from the metadata is only a temporary
+        // thing until metadata is immutable. At that point, current ensemble
+        // becomes a property of the LedgerHandle itself.
```

Comment:
I know `LedgerHandle` is the mix of write and read handle. However this method is exposed to ReadOnlyHandle. It is a bit confused when we are saying `current ensemble` for a readonly handle. so I would suggest adding more comments here to clarify the meanings.

Reply:
There'll be a big comment on the ReadOnlyLedgerHandle implementation. The whole class hierarchy here is broken. LedgerHandle should derive from ReadOnlyLedgerHandle, not vice versa. ReadOnlyLedgerHandle shouldn't do recovery. Recovery should be a separate operation that occurs before the ledger handle is even created. However, currently this is a big change because of the lack of separation between the *Op classes and LedgerHandle. This getCurrentEnsemble() is the ugliness I was referring to on dev@ when I said the solution isn't nice. The add op has to call back to the ledger handle to get the new ensemble after a failure. It would be better if, when an ensemble change occurs, all outstanding pending add ops are passed the new ensemble **by** the ledger handle. I might actually change it do that. The ops shouldn't depend on LedgerHandle. The objects they require should be passed in as parameters. There's so much janky stuff in this code, it can be painful to not fix it all at once.

Reply:
@ivankelly we all understand the problem and motivation behind this. however that's a different topic to address all these. so let's not try to couple that here. 

Here what I am asking is adding a comment to make thing clear. That's it. I want people who might be working on this file or when people look back at the git commit, we know exactly what was the thought of `getCurrentEnsemble`.